docs/src/tutorial/pure.rst

```diff
@@ -302,6 +273,49 @@ can be augmented with the following :file:`.pxd` file :file:`dostuff.pxd`:
 The :func:`cython.declare()` function can be used to specify types for global
 variables in the augmenting :file:`.pxd` file.
 
+.. _annotations:
+
+Annotations
+-----------
+
+Python annotations can be used to declare argument types, as shown in the
+following example.  To avoid conflicts with other kinds of annotation
+usages, this can be disabled with the directive ``annotation_typing=False``.
+
+  .. literalinclude:: ../../examples/tutorial/pure/annotations.py
+
+Note the use of ``cython.int`` rather than ``int`` - Cython does not translate
+an ``int`` annotation to a C integer by default since the behaviour can be
+quite different with respect to overflow and division.
+
+Annotations can be combined with the ``@cython.exceptval()`` decorator for non-Python
+return types:
+
+  .. literalinclude:: ../../examples/tutorial/pure/exceptval.py
+
+Note that the default exception handling behaviour when returning C numeric types
+is to check for ``-1``, and if that was returned, check Python's error indicator
+for an exception.  This means, if no ``@exceptval`` decorator is provided, and the
+return type is a numeric type, then the default with type annotations is
+``@exceptval(-1, check=True)``, in order to make sure that exceptions are correctly
+and efficiently reported to the caller.  Exception propagation can be disabled
+explicitly with ``@exceptval(check=False)``, in which case any Python exceptions
+raised inside of the function will be printed and ignored.
+
+Since version 0.27, Cython also supports the variable annotations defined
+in `PEP 526 <https://www.python.org/dev/peps/pep-0526/>`_. This allows to
+declare types of variables in a Python 3.6 compatible way as follows:
+
+.. literalinclude:: ../../examples/tutorial/pure/pep_526.py
+
+There is currently no way to express the visibility of object attributes.
+
+Cython does not support the full range of annotations described by PEP-0526.
+For example it does not understand features from the ``typing`` module such 
+as ``Optional[]`` or typed containers such as ``List[str]``. Such annotations
+typically do not provide information that is useful to Cython in generating
+efficient C code.
```

Comment:
It's true that these type hints do not always help, but they _could_ in many situations, even if it's just the hint that it's "most likely" a dict (or a list) that will get passed, and never a tuple or other container.
```suggestion
Cython does not support the full range of annotations described by PEP-484.
For example it does not currently understand features from the ``typing`` module
such  as ``Optional[]`` or typed containers such as ``List[str]``. This is partly
because some of these type hints are not relevant for the compilation to
efficient C code. In other cases, however, where the generated C code could
benefit from these type hints but does not currently, help is welcome to
improve the type analysis in Cython.
```